src/app-layer-dnp3.c

```diff
@@ -0,0 +1,2616 @@
+        if (unlikely(tmp == NULL)) {
+            return 0;
+        }
+        buffer->buffer = tmp;
+        buffer->size = buffer->len + len;
+    }
+    memcpy(buffer->buffer + buffer->len, buf, len);
+    buffer->len += len;
+
+    return 1;
+}
+
+/**
+ * DNP3 buffer helper function - pulls up the data to the beginning of the
+ * buffer to make room at the end for more data.
+ */
+static void DNP3BufferTrim(DNP3Buffer *buffer)
+{
+    if (buffer->offset == buffer->len) {
+        DNP3BufferReset(buffer);
+    }
+    else if (buffer->offset > 0) {
+        memmove(buffer->buffer, buffer->buffer + buffer->offset,
+            buffer->len - buffer->offset);
+        buffer->len = buffer->len - buffer->offset;
+        buffer->offset = 0;
+    }
+}
+
+/**
+ * \brief Decode DNP3 application objects.
+ *
+ * This function decoded known DNP3 application objects. As the
+ * protocol isn't self describing, we can only decode the buffer while
+ * the application objects are known.  As soon as an unknown
+ * group/variation is hit, we must stop processing.
+ *
+ * \param buf the input buffer
+ * \param len length of the input buffer
+ * \param objects pointer to list where decoded objects will be stored.
+ *
+ * \retval 1 if all objects decoded, 0 if all objects could not be decoded (
+ *    unknown group/variations)
+ */
+static int DNP3DecodeApplicationObjects(uint8_t *buf, uint32_t len,
+    DNP3ObjectList *objects)
+{
+    int retval = 0;
+
+    if (buf == NULL || len == 0) {
+        return 1;
+    }
+
+    while (len) {
+        uint32_t offset = 0;
+
+        if (len < sizeof(DNP3ObjHeader)) {
+            SCLogInfo(
+                "Insufficient data: not enough data for object header (%d).",
+                len);
+            SCLogInfo("%02x", buf[2]);
+            goto done;
+        }
+        DNP3ObjHeader *header = (DNP3ObjHeader *)buf;
+        offset += sizeof(DNP3ObjHeader);
+
+        DNP3Object *object = SCCalloc(1, sizeof(*object));
+        if (object != NULL) {
+            object->group = header->group;
+            object->variation = header->variation;
+            TAILQ_INSERT_TAIL(objects, object, next);
+        }
+
+        SCLogDebug("Group: 0x%02x (%d); Variation: 0x%02x (%d); "
+            "Qualifier: 0x%02x (%d)",
+            header->group, header->group,
+            header->variation, header->variation,
+            header->qualifier, header->qualifier);
+        uint8_t prefix_code = DNP3_OBJ_PREFIX(header->qualifier);
+        uint8_t range_code = DNP3_OBJ_RANGE(header->qualifier);
+        object->prefix = DNP3_OBJ_PREFIX(header->qualifier);
+        object->range = DNP3_OBJ_RANGE(header->qualifier);
+
+        int prefix_len = DNP3_OBJ_PREFIX_CODE_MAP[prefix_code];
+        if (prefix_len < 0) {
+            goto unknown;
+        }
+
+        uint32_t count = 0;
+
+        /* IEEE 1815-2012, Table 4-5. */
+        switch (range_code) {
+            case 0x00: {
```

Comment:
Normally I'd say yes, but even in the spec, the hex value is the "name" of the field, unless we wanted to do something like RANGE_CODE_1, ..., RANGE_CODE_9?